### ISSUE TITLE
Serialize Uint8Arrays better

### ISSUE THREAD

**Issue by benjamingr** [puppeteer/puppeteer#2427]
### Steps to reproduce

**Tell us about your environment:**

* Puppeteer version: 1.3.0
* Platform / OS version: macOS Sierra
* URLs (if applicable): ...
* Node.js version: master

**What steps will reproduce the problem?**

_Please include code that reproduces the issue._

1. Open a new page with puppeteer
2. Run `page.evaluate(() => new Uint8Array(1024 * 1024 * 10)`

**What is the expected result?**

Puppeteer takes over 20 seconds to serialize/deserialize 10MB and the resulting object is not a Uint8Array. 

**What happens instead?**

A Uint8Array should be returned and it should take a reasonable amount of time (single-digit miliseconds tops and not seconds).

----

This basically would require inspecting objects returned from `evaluate`, detecting they're Uint8Arrays and then serializing them differently.

**Comment by aslushnikov** [comment]
@benjamingr pptr relies on devtools protocol for object passing; devtools protocol is basically a JRPC, so there's obviously a lot of overhead when passing large objects. 

The JRPC part: There's been some discussions on binary format, but I don't think there's any plan to introduce anything like this. Even though, passing 10mb from one process to another without shared memory will take much more then ~millisecond.

The "shared memory" is not an option in our current architecture and is not a priority for out main scenarios (e.g. "remote debugging").

For your usecase, I'd recommend using `page.evaluateHandle` to avoid serializing altogether. If you definitely need to pass massive chunks of data from page to pptr, I'd try rolling custom serializer with compression. For example, large-dimensional sparse arrays (like in your example) could be encoded quite effectively.

**Comment by benjamingr** [comment]
@aslushnikov please consider reopening this issue.

> @benjamingr pptr relies on devtools protocol for object passing; devtools protocol is basically a JRPC, so there's obviously a lot of overhead when passing large objects.

That's an implementation detail puppeteer can overcome on its side. For example by serializing the object on one end - feeding it to JRPC and then deserializing it at the other end.

My use case for moving a lot of data is very real, it's not farfetched at all and I believe it's pretty common. Plus if you're supporting downloading files - moving buffers that contain files only makes sense.

> The "shared memory" is not an option in our current architecture and is not a priority for out main scenarios (e.g. "remote debugging").

I didn't suggest that given how isolates work - though it's actually quite possible in this particular case since it's not structured data. I wouldn't go that route though since it'd require actual work from V8.

> For your usecase, I'd recommend using page.evaluateHandle to avoid serializing altogether.

I haven't told you what my use case is. We haven't even discussed use cases. `page.evaluateHandle` wouldn't work for it. I need to move about ~5MB of data 3 times a second (from WebRTC if it matters) from Chrome to Node. 

> I'd try rolling custom serializer with compression.

I realize I can do that. What I currently do is use `TextEncoder/TextDecoder` to serialize the data as a single char-per-byte string (using windows-1252 encoding) and deserialize it in Node on the other side which is fast (since passing strings is much faster). 

Puppeteer can implement a similar serialize/deserialize strategy which would make passing typed arrays much faster. I think passing native JavaScript types should generally be supported.

**Comment by aslushnikov** [comment]
> @aslushnikov please consider reopening this issue.

Sure!

> My use case for moving a lot of data is very real, it's not farfetched at all and I believe it's pretty common. 

Feel free to send a PR

**Comment by cancerberoSgx** [comment]
@benjamingr - I tried using TextEncoder with the encoding you mentioned and others but it won't work for me, although I realized passing the data as string is much faster that my previous approach that was passing as number[] Array.from(arrayBufferView) <-> new UInt8Array(numberArray). 

What did work for me was using Node.js Buffers and the library 'buffer' (https://github.com/feross/buffer)  in the browser: 

```js
// load 'buffer' library. It will be available at window.buffer.Buffer.
await this.page!.addScriptTag({content:  readFileSync('test/assets/buffer-5.4.3.min.js').toString() })

// store a file as a Buffer in window.fileContents3 global variable 
// (simulates the image array buffer view extracted from canvas)
await this.page.evaluate((s) =>
  window.fileContents3 = window.buffer.Buffer.from(s, 'binary')
  , readFileSync('test.mov').toString('binary'))

// get data as Node.js Buffer
const fileContents = Buffer.from(await this.page.evaluate(() =>
  window.fileContents3.toString('binary')
), 'binary')
console.timeEnd('c')

writeFileSync('tmp_test.mov', fileContents)
```

Numbers aprox are : 
 * 1.5 mb jpeg : 300ms
 * 13mb mov : 2600ms

(More or less a linear relationship between time+size?)

Although I will keep researching around this, these are some Notes / questions I have. In my particular problem which is passing a HTMLCanvasElement image (both supporting ImageData and encoded images as jpeg, png):

 * It would make any difference to use ` page.exposeFunction()` instead page.evaluate() here in terms of performance ? 
 * I didn't want to use asDataUrl since string encoding sounds extra overhead that I don't need, but I will give it a try.
 * Is it possible to use a node.js Worker consumable from the browser ? (or access a socket using devtools protocols / APIs) ? 
 + or in general, is it possible to share memory somehow (not copying) perhaps using a devtools protocol API or just posting the data to a local named pipe or net.Server listening at some local addrress/port ? (I know the later is possible but I wonder if it's worthwhile to try)

BTW I'm building a camera capture node.js library / CLI https://www.npmjs.com/package/camera-capture which is something that currently doesn't exists in npm or very hard to install - and puppeteer makes this job pretty easy . Optionally I can record the video using MediaRecorder (or desktop / audio) .... so so far I'm happy besides this little speed detail . 

In my case the goal is to pass frames to node.js as fast as possible so they can be processed and rendered "real-time" using native libraries in native surfaces (let's say opencv & cairo based application). My best times are 480x320 20fps (but encoded as jpg / webp while Ideally I would like to have the raw image data since the app doesn't need to decode each frame for processing / rendering. 

Thanks

**Comment by benjamingr** [comment]
What I ended up doing is either using the filesystem for this or a Node HTTP server I POST to.

**Comment by cancerberoSgx** [comment]
yeap I was thinking on posting to a named pipe or socket and see what happen - but that will complicate code so it must really worth it.  I'm also tempted to postMessage to a native Node.js Worker but I now It won't work just because both things are call workers :) 

I was looking at the sources and the implementation for this issue seems to be not hard - assuming the changes are only in the .js files. since I'm not familiar with the code I wonder if it also should imply modifying C++ ?  thanks  And if designed I think the proper structure to transfer should be ArrayBuffer and not ArrayBufferView what do you say?

Finally I was thinking about return String.fromCharCode.apply(null, new Uint8Array(buffer)); not about using it since it fails for large arrays, but made me think about the fact that the Uint8Array spreads like number[] arg in that call - I wonder If puppeteer + exposeFunction cannot be cheated in a similar fashion - perhaps sing spread args

**Comment by benjamingr** [comment]
It's actually pretty simple and fast with http:

```js
          const server = createServer((req, res) => {
            if (req.url.includes('session')) { // 
              res.setHeader('Access-Control-Allow-Origin', '*');
              // can parse the body stream here easily and read data from the browser
              res.end(JSON.stringify(DATA_I_WANT_TO_SEND_BACK_FROM_NODE));
              server.close();
            }
          });
          await new Promise(resolve => server.listen(0, resolve));
          const result = await page.evaluate(async (port) => {
            const session = await fetch(`http://localhost:${port}/session`).then(x => x.json());
            // session now available here, we passed a lot of data to puppeteer
            // the other direction is just POSTing it
            return doSomeAction(session);
          }, server.address().port);
```

The single-width-byte encoding trick was mostly useful when that was hard.

**Comment by cancerberoSgx** [comment]
aja I already have an http server for the index.html -   my problem is passing data form browser to node

I will give a try also using websockets and another experiment I will perform  sharing a webrtc session between node and browser which I think is totally possible and probably optimal for streaming video from pptr to node.js...  in my case (camera capture) that could imply not read/write image data at all and just read the stream in node, If I make it work I will update here. thanks.

**Comment by cancerberoSgx** [comment]
> ```js
> res.end(JSON.stringify(DATA_I_WANT_TO_SEND_BACK_FROM_NOD
> ```

why are you using JSON - I I though you were dealing with array buffer and binary data iin which case could you just write the buffer directly and consume it as array buffer in the browser ? (perhaps you need to use the correct  content type in the response ? I don't think JSON stringify of 10mb  will never be fast or am I missing something ? .... ?

**Comment by benjamingr** [comment]
> why are you using JSON - I I though you were dealing with array buffer and binary data iin which case could you just write the buffer directly and consume it as array buffer in the browser ?

Yes of course I was just giving an example - my real code writes a node stream and reads it as a stream on the other side using `fetch` and even backpressure works (I process as quickly as I can on the client and `read`).

**Comment by pke** [comment]
[This method](https://github.com/GoogleChrome/puppeteer/issues/2427#issuecomment-536002538) does not seem to work with services like browserless.io? The connection is always dropped.

Also the local http server cannot be used with the online service, can it?

**Comment by jimmywarting** [comment]
I agree that there needs to be a better way of transfering streams and large bytes of data.
The idea to be able to transfer a ReadableStream over postMessage or something would be sweet now that nodejs now also has whatwg streams. Even chrome supports transferable streams between workers. puppeteer should be able to do something similar i think 
The solution would have to be for puppeteer two establish a websocket pair between node and the browser.


```js
// Some wishful programming:
const rs = new ReadableStream({
 start(ctrl) {
   ctrl.enqueue(new Uint8Array[97, 98, 99])
   ctrl.close()
 }
})

page.evaluate(rs => {
  const ws = new WritableStream({
    write(chunk) {
      console.log(chunk)
    }
  })
  
  rs.pipeTo(ws)
}, rs)
```
what this ☝️ needs to to in the background is to establish a binary websocket connection behind the scenes

**Comment by louzhedong** [comment]
I'm interested in taking a narrow pass at this if it's still open. I'm thinking specifically about preserving typed-array shape and improving the transfer path for `Uint8Array` results from `evaluate`, with focused benchmarks/tests rather than trying to redesign large-object transfer more broadly.

**Comment by Develop-KIM** [comment]
Hi! I'd like to work on this issue. @louzhedong expressed interest last month — are you
still working on it? Happy to defer or collaborate if so.

Before writing code, I'd like to get maintainer feedback on the approach, since this
touches evaluate semantics.

**Current behavior** (puppeteer 24.43.1, Chrome 148, M-series Mac):

| Size | `evaluate()` today | base64 round-trip |
|------|-------------------|-------------------|
| 1MB  | 341ms, returns plain `Object` (`{"0":…,"1":…}`) | 25ms, intact `Uint8Array` |
| 10MB | 4.2s, returns plain `Object` | 302ms, intact `Uint8Array` |

**Root cause**: on the CDP path, `Runtime.callFunctionOn` with `returnByValue: true`
JSON-serializes inside Chromium, so a `Uint8Array` becomes an index-keyed object before
puppeteer ever sees it. On the BiDi path it's worse `BidiDeserializer` has no case for
`typedarray`, so `evaluate` returns `undefined`.

**Proposed design**:

1. Wrap the user's page function so that if the top-level return value is a typed array
   (`ArrayBuffer.isView`) or `ArrayBuffer`, it is encoded in-page to base64 — using native
   `Uint8Array.prototype.toBase64()` where available, with a chunked `btoa` fallback
   and returned as a small marker object.
2. On the Node side, detect the marker and decode via the existing
   `stringToTypedArray()` in `src/util/encoding.ts` (which already prefers
   `Uint8Array.fromBase64`), reconstructing the original typed-array subclass.
3. Apply to both the CDP and BiDi evaluate paths, with regression tests covering type
   fidelity and a size benchmark.

**Open questions for maintainers**:

- Should this be the default behavior or opt-in? Returning a real `Uint8Array` where users
  previously got an index-keyed object is technically a behavior change (though arguably
  a bug fix, and consistent with the Buffer→Uint8Array direction of #12823).
- Scope: top-level return values only for a first pass, or also typed arrays nested inside
  returned objects/arrays?
- Should `ArrayBuffer`/`DataView` be included, or `Uint8Array`/typed arrays only?

If this direction sounds acceptable I'll put together a PR.
